api/src/main/java/com/github/retrooper/packetevents/protocol/player/storage/FastUserStorage.java

```diff
@@ -0,0 +1,112 @@
+package com.github.retrooper.packetevents.protocol.player.storage;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+/**
+ * Fast User Storage - An API usage class
+ * This class provides efficient runtime value association with the User class (the association exists for as along as the User doesn't quit - it's non-permanent).
+ * It can be used for many different purposes, such as, but not limited to:
+ * 1. Storing the plugin's packet handler for the User even before UUID is known. This could be achieved for example via:
+ *
+ * <pre> {@code
+ *  private final TypedStorageValueId<MyUser> identifier = TypedStorageValueId.identifierFor(this.myPluginInstance);
+ *
+ *  @Override
+ *  public void onUserConnect(UserConnectEvent event) {
+ *      event.getUser().storeRuntime(this.identifier, new MyUser());//associate
+ *  }
+ *
+ *  @Override
+ *  public void onPacketReceive(PacketReceiveEvent event) {
+ *      MyUser user = event.getUser().getStored(this.identifier);
+ *      user.packetHandler().handle(event);
+ *  }
+ * }</pre>
+ * <p>
+ * 2. Replacing a Map - Even tho Maps are extremely fast, their performance can slightly decrease with size. This Storage class offers constant O(1)
+ * performance, independently to the amount of Users or the length of the data array. With 100 users, the performance of this class was often found to be
+ * about ~35% faster than that of a UUID Map.
+ *
+ * @author ShadowOfHeaven
+ */
+
+public final class FastUserStorage {
+
+    //Ensure the array's element thread visibility by using MethodHandles
+    //It was found to be slightly faster than VarHandles with excessive amount of calls
+    private static final MethodHandle setElement = MethodHandles.arrayElementSetter(Object[].class);//array, index, value
+    private static final MethodHandle getElement = MethodHandles.arrayElementGetter(Object[].class);//array, index
+    private static final Map<Object, Integer> pluginToId = new ConcurrentHashMap<>(2);
```

Comment:
if the keys here are plugin instances, this should definetely be a `WeakHashMap`
otherwhise this will cause memory leaks when a plugin gets disabled or unloaded

because of this you will need to do manual synchronization, see some coment below for mor info

Reply:
im also not sure if its a good idea to key these ids by "plugin" instance - in my opinion it makes more sense to key the id map by the class of the identifier type, so you can create multiple keys per plugin

Reply:
> if the keys here are plugin instances, this should definetely be a `WeakHashMap` otherwhise this will cause memory leaks when a plugin gets disabled or unloaded
> 
> because of this you will need to do manual synchronization, see some coment below for mor info

Guava's MapMaker is the best solution then

Reply:
> im also not sure if its a good idea to key these ids by "plugin" instance - in my opinion it makes more sense to key the id map by the class of the identifier type, so you can create multiple keys per plugin

The main idea behind this is that you create one key, which holds one custom object, that can encapsulate all the other values associated. Otherwise this could lead to memory being wasted